extensions/funqy/funqy-knative-events/runtime/src/main/java/io/quarkus/funqy/knative/events/CloudEvent.java

```diff
@@ -1,19 +1,32 @@
 package io.quarkus.funqy.knative.events;
 
+import java.net.URI;
 import java.time.OffsetDateTime;
+import java.util.Iterator;
+import java.util.Map;
 
 /**
- * Cloud event. Represents only the headers. No data.
+ * CloudEvent.
  *
  */
-public interface CloudEvent {
+public interface CloudEvent<T> {
+
     String id();
 
     String specVersion();
 
-    String source();
+    URI source();
```

Comment:
I think URI for source is a bit to strict as it does not have to be URI, from the spec

> The application MAY use UUIDs, URNs, DNS authorities or an application-specific scheme to create unique source identifiers.

and examples:
````
Application-specific identifiers
/cloudevents/spec/pull/123
/sensors/tn-1234567/alerts
1-555-123-4567
````

so enforcing it on the type of cloud event might be problematic.

Reply:
For fine grain performance using strings is better, but are you sure URI isn't required? Spec says source is a "URI-reference"

https://github.com/cloudevents/spec/blob/master/spec.md

Reply:
based description and examples it can be non URI even though it's recommended to be an URI. My point is that we don't have to enforce on the api as it is mainly for serialisation.

Reply:
Usage of `URI` definitely incurs performance drop.
Last time I tried to benchmark `Funqy vs. CNCF cloudevent java sdk` non trivial difference was caused by `ce-sdk` using `URI`

I picked it primarily because https://github.com/cloudevents/sdk-java/tree/master/api does use `URI`. I wish we were close to their API in case we wanted to adopt it.

Reply:
personally having it as string offloads Funqy from doing heavy work and of someone wants to use URI can still to the toString on function side.